crates/transpiler/src/passes/litinski_transformation.rs

```diff
@@ -0,0 +1,218 @@
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::imports::PAULI_EVOLUTION_GATE;
+use qiskit_circuit::operations::{
+    multiply_param, Operation, OperationRef, Param, PyGate, StandardGate,
+};
+use qiskit_circuit::packed_instruction::PackedInstruction;
+use qiskit_circuit::{Clbit, Qubit, VarsMode};
+
+use qiskit_quantum_info::clifford::Clifford;
+use qiskit_quantum_info::sparse_observable::PySparseObservable;
+
+use smallvec::smallvec;
+use std::f64::consts::PI;
+
+use crate::TranspilerError;
+
+// List of gate names supported by the pass: the pass is skipped if the circuit
+// contains gate names outside of this list.
+const SUPPORTED_GATE_NAMES: &[&str; 19] = &[
+    "id", "x", "y", "z", "h", "s", "sdg", "sx", "sxdg", "cx", "cz", "cy", "swap", "iswap", "ecr",
+    "dcx", "t", "tdg", "rz",
+];
+
+// List of rotation gate names: the pass is skipped if the circuit contains
+// no gate names in this list.
+const ROTATION_GATE_NAMES: &[&str; 3] = &["t", "tdg", "rz"];
+
+/// Expresses a given circuit as a sequence of Pauli rotations followed by a final Clifford operator.
+/// Returns the list of rotations in the sparse format: (sign, paulis, indices).
+pub fn extract_rotations(
+    circuit: &[(String, Vec<usize>)],
+    nqubits: usize,
+) -> Vec<(bool, String, Vec<u32>)> {
+    let mut clifford = Clifford::identity(nqubits);
+    let mut rotations: Vec<(bool, String, Vec<u32>)> = Vec::new();
+
+    for (gate_name, qbits) in circuit.iter() {
+        match gate_name.as_str() {
+            "id" => {}
+            "x" => clifford.append_x(qbits[0]),
+            "y" => clifford.append_y(qbits[0]),
+            "z" => clifford.append_z(qbits[0]),
+            "h" => clifford.append_h(qbits[0]),
+            "s" => clifford.append_s(qbits[0]),
+            "sdg" => clifford.append_sdg(qbits[0]),
+            "sx" => clifford.append_sx(qbits[0]),
+            "sxdg" => clifford.append_sxdg(qbits[0]),
+            "cx" => clifford.append_cx(qbits[0], qbits[1]),
+            "cz" => clifford.append_cz(qbits[0], qbits[1]),
+            "cy" => clifford.append_cy(qbits[0], qbits[1]),
+            "swap" => clifford.append_swap(qbits[0], qbits[1]),
+            "iswap" => clifford.append_iswap(qbits[0], qbits[1]),
+            "ecr" => clifford.append_ecr(qbits[0], qbits[1]),
+            "dcx" => clifford.append_dcx(qbits[0], qbits[1]),
+            "rz" => {
+                rotations.push(clifford.get_inverse_z(qbits[0]));
+            }
+            _ => panic!("Unsupported gate {}", gate_name),
+        }
+    }
+
+    rotations
+}
+
+#[pyfunction]
+#[pyo3(name = "run", signature = (dag, fix_clifford=true))]
+pub fn run_litinski_transformation(
+    py: Python,
+    dag: &mut DAGCircuit,
```

Comment:
From a quick check it looks like we can remove `mut` here, right?
```suggestion
    dag: &DAGCircuit,
```

Reply:
fixed